loyalty_outlier_analysis.ipynb

```diff
@@ -0,0 +1,476 @@
+    "    print(pd.DataFrame(C, index=cm_row_labels, columns=cm_column_labels))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train_and_validation_df.dtypes"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The decision tree classifier likes only numeric features, so we need to remove all the non-numeric features (including datetime)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tree_df = train_and_validation_df[train_and_validation_df.columns.difference(['first_active_month',\n",
+    "                                                                              'target',\n",
+    "                                                                              'authorized_flag_top',\n",
+    "                                                                              'category_1_top',\n",
+    "                                                                              'outlier'\n",
+    "                                                                             ])]\n",
+    "tree_df.fillna(tree_df.mean(), inplace=True)\n",
+    "data_train, data_test, label_train, label_test = train_test_split(tree_df, train_and_validation_df['outlier'], test_size=0.2)\n",
+    "tree_df.head()\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Fiddling with the `max_depth` value gives more or less granularity to the decision tree. It doesn't take very long to build even a deep tree."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "clf = DecisionTreeClassifier(max_depth=9)\n",
+    "clf.fit(data_train, label_train)\n",
+    "\n",
+    "draw_decision_tree(clf, data_train.columns.values)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "evaluate_classification_results(clf, data_train, label_train)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "evaluate_classification_results(clf, data_test, label_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The result is that a decision tree is not very good. In general, decision trees are probably a bad idea to predict rare events. By increasing the depth of the tree, we can get good-ish results on the training set, but the tree is horribly overfitted and produces bad results on the validation set."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Penalizing Outlier Prediction Mistakes\n",
+    "\n",
+    "There are [techniques for handling imbalanced classes](https://elitedatascience.com/imbalanced-classes). Let's\n",
```

Comment:
Would be great if you could explain these techniques in brief on Thursday – I think it's educative to compare different means of achieving the same thing like this. :-)

Reply:
Sure, I'll do that. I'm booked the whole afternoon this Thursday for my new project, but there will be other chances.

Reply:
Cool.

So maybe we skip this week since you can't and Boris is in Barcelona? And we continue next week Tuesday or something (2 weeks before the deadline).